### PR TITLE
Rework free list to use less memory.

### PR DESCRIPTION
**Pull request by zherczeg**
This patch is a follow-up for #1835. It can be landed independently or merged with the other PR.

### CODE REVIEW

**Comment by sbc100** [comment]
Do you have some way of measuring the impact of this change?   For example, are you seeing this improve memory usage in some real world use case that you have?

**Comment by zherczeg** [comment]
I have made a measurement using the freya tool for valgrind. I have modified the interpreter to run all test files in a single loop, and grouped the data using the following config file:

```
# discarded allocations
-                  /usr/include/c++/9/bits
-                  /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.28
-                  m_replacemalloc/vg_replace_malloc.c
-                  (mmap)

# Interpreter components
interp             wabt/src/interp
interp-tool        wabt/src/tools
interp-util        wabt/src

[Interpreter]
  (interp)
  (interp-tool)
  (interp-util)

# Everything else
[other] 1 +
```

The results are here:
https://gist.github.com/zherczeg/77df5696b48b92f8e51852bf45d104fa

FreeList does not seems to be a major memory consumer, Peak looks like 1K. It seems the default consumption for Memory::Memory, frames_ and values_ in Thread are 64K, and those are much bigger (3 * 64K).

**Comment by zherczeg** [comment]
@sbc100 Is this measurement data ok, or do you need something else?

File: src/interp/interp-inl.h [hunk 1]
```diff
@@ -139,95 +139,107 @@ inline Frame::Frame(Ref func,
...
 
 template <typename T>
-auto FreeList<T>::count() const -> Index {
-  return list_.size() - free_.size();
+FreeList<T>::~FreeList<T>() {
+  for (auto object : list_) {
+    if ((reinterpret_cast<uintptr_t>(object) & ptrFreeBit) == 0) {
+      delete object;
+    }
+  }
```
Comment (sbc100) [inline src/interp/interp-inl.h:189]: Where these just leaked before?

**Review by sbc100 — APPROVED** [review]
Looks good!

File: src/interp/interp-inl.h (hunk 1 above)
Comment (zherczeg) [reply src/interp/interp-inl.h:189]: It was a `std::unique_ptr` before, but casting a struct into a number is not necessary working as expected (implementation dependent).

**Comment by sbc100** [comment]
I think this change might be responsible a crash we are seeing.  ASAN picks it up:

```
Running third_party/halide/halide/test/correctness_align_bounds_wasm_impl expect=success with HL_TARGET=wasm-32-wasmrt with HL_JIT_TARGET=wasm-32-wasmrt
==5569==WARNING: ASan doesn't fully support makecontext/swapcontext functions and may produce false positives in some cases!
=================================================================
==5569==ERROR: AddressSanitizer: heap-use-after-free on address 0x6060003658a0 at pc 0x56275c3a1258 bp 0x7ffc0f408d10 sp 0x7ffc0f408d08
READ of size 8 at 0x6060003658a0 thread T0
    #0 0x56275c3a1257 in wabt::interp::FreeList<wabt::interp::Object*>::~FreeList() third_party/wabt/src/interp/interp-inl.h:187:7
    #1 0x56275c39f91f in wabt::interp::Store::~Store() third_party/wabt/src/interp/interp.h:444:7
    #2 0x56275c3a0483 in Halide::Internal::WasmModuleContents::~WasmModuleContents() third_party/halide/halide/src/WasmExecutor.cpp:2287:35
    #3 0x56275c398992 in void Halide::Internal::destroy<Halide::Internal::WasmModuleContents>(Halide::Internal::WasmModuleContents const*) third_party/halide/halide/src/WasmExecutor.cpp:2696:5
    #4 0x56275b48337c in decref third_party/halide/halide/src/IntrusivePtr.h:85:17
    #5 0x56275b48337c in ~IntrusivePtr third_party/halide/halide/src/IntrusivePtr.h:113:9
    #6 0x56275b48337c in ~WasmModule third_party/halide/halide/src/WasmExecutor.h:27:8
    #7 0x56275b48337c in Halide::PipelineContents::invalidate_cache() third_party/halide/halide/src/Pipeline.cpp:91:9
    #8 0x56275b483528 in Halide::PipelineContents::clear_custom_lowering_passes() third_party/halide/halide/src/Pipeline.cpp:134:9
    #9 0x56275b4813a8 in Halide::PipelineContents::~PipelineContents() third_party/halide/halide/src/Pipeline.cpp:130:9
    #10 0x56275b464ef2 in void Halide::Internal::destroy<Halide::PipelineContents>(Halide::PipelineContents const*) third_party/halide/halide/src/Pipeline.cpp:152:5
    #11 0x56275ab15557 in decref blaze-out/k8-asan-fastbuild/genfiles/third_party/halide/halide/src/halide_uppercase/Halide.h:2199:17
    #12 0x56275ab15557 in ~IntrusivePtr blaze-out/k8-asan-fastbuild/genfiles/third_party/halide/halide/src/halide_uppercase/Halide.h:2227:9
    #13 0x56275ab15557 in ~Pipeline blaze-out/k8-asan-fastbuild/genfiles/third_party/halide/halide/src/halide_uppercase/Halide.h:15309:7
    #14 0x56275ab15557 in ~Func blaze-out/k8-asan-fastbuild/genfiles/third_party/halide/halide/src/halide_uppercase/Halide.h:17106:7
    #15 0x56275ab15557 in main third_party/halide/halide/test/correctness/align_bounds.cpp:86:5
    #16 0x7f4176b038d2 in __libc_start_main (/usr/grte/v5/lib64/libc.so.6+0x628d2) (BuildId: 7cfed7708e5ab7fcb286b373de21ee76)
    #17 0x56275aa47ae9 in _start /usr/grte/v5/debug-src/src/csu/../sysdeps/x86_64/start.S:120

0x6060003658a0 is located 0 bytes inside of 64-byte region [0x6060003658a0,0x6060003658e0)
freed by thread T0 here:
    #0 0x56275ab105a2 in operator delete(void*, unsigned long) third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:164:3
    #1 0x5627652976d1 in wabt::interp::Object::~Object() third_party/wabt/src/interp/interp.cc:313:19
    #2 0x56275c3a1193 in wabt::interp::FreeList<wabt::interp::Object*>::~FreeList() third_party/wabt/src/interp/interp-inl.h:187:7
    #3 0x56275c39f91f in wabt::interp::Store::~Store() third_party/wabt/src/interp/interp.h:444:7
    #4 0x56275c38d1ea in Halide::Internal::WasmModuleContents::WasmModuleContents(Halide::Module const&, std::__u::vector<Halide::Argument, std::__u::allocator<Halide::Argument> > const&, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, std::__u::map<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> >, Halide::JITExtern, std::__u::less<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > >, std::__u::allocator<std::__u::pair<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const, Halide::JITExtern> > > const&, std::__u::vector<Halide::Internal::JITModule, std::__u::allocator<Halide::Internal::JITModule> > const&) third_party/halide/halide/src/WasmExecutor.cpp:2312:5
    #5 0x56275c398a62 in Halide::Internal::WasmModule::compile(Halide::Module const&, std::__u::vector<Halide::Argument, std::__u::allocator<Halide::Argument> > const&, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, std::__u::map<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> >, Halide::JITExtern, std::__u::less<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > >, std::__u::allocator<std::__u::pair<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const, Halide::JITExtern> > > const&, std::__u::vector<Halide::Internal::JITModule, std::__u::allocator<Halide::Internal::JITModule> > const&) third_party/halide/halide/src/WasmExecutor.cpp:2718:32
    #6 0x56275b47238c in Halide::Pipeline::compile_jit(Halide::Target const&) third_party/halide/halide/src/Pipeline.cpp:635:33
    #7 0x56275b4786d6 in Halide::Pipeline::realize(Halide::JITUserContext*, Halide::Pipeline::RealizationArg, Halide::Target const&, Halide::ParamMap const&) third_party/halide/halide/src/Pipeline.cpp:1130:5
    #8 0x56275b477724 in Halide::Pipeline::realize(Halide::JITUserContext*, std::__u::vector<int, std::__u::allocator<int> >, Halide::Target const&, Halide::ParamMap const&) third_party/halide/halide/src/Pipeline.cpp:757:9
    #9 0x56275b476e8e in Halide::Pipeline::realize(std::__u::vector<int, std::__u::allocator<int> >, Halide::Target const&, Halide::ParamMap const&) third_party/halide/halide/src/Pipeline.cpp:738:12
    #10 0x56275aba88d0 in Halide::Func::realize(std::__u::vector<int, std::__u::allocator<int> >, Halide::Target const&, Halide::ParamMap const&) third_party/halide/halide/src/Func.cpp:3276:23
    #11 0x56275ab14b3f in main third_party/halide/halide/test/correctness/align_bounds.cpp:52:32
    #12 0x7f4176b038d2 in __libc_start_main (/usr/grte/v5/lib64/libc.so.6+0x628d2) (BuildId: 7cfed7708e5ab7fcb286b373de21ee76)
    #13 0x56275aa47ae9 in _start /usr/grte/v5/debug-src/src/csu/../sysdeps/x86_64/start.S:120

previously allocated by thread T0 here:
    #0 0x56275ab0f91d in operator new(unsigned long) third_party/llvm/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:95:3
    #1 0x562765295da7 in wabt::interp::Store::Store(wabt::Features const&) third_party/wabt/src/interp/interp.cc:203:24
    #2 0x56275c38d1d0 in Halide::Internal::WasmModuleContents::WasmModuleContents(Halide::Module const&, std::__u::vector<Halide::Argument, std::__u::allocator<Halide::Argument> > const&, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, std::__u::map<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> >, Halide::JITExtern, std::__u::less<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > >, std::__u::allocator<std::__u::pair<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const, Halide::JITExtern> > > const&, std::__u::vector<Halide::Internal::JITModule, std::__u::allocator<Halide::Internal::JITModule> > const&) third_party/halide/halide/src/WasmExecutor.cpp:2312:13
    #3 0x56275c398a62 in Halide::Internal::WasmModule::compile(Halide::Module const&, std::__u::vector<Halide::Argument, std::__u::allocator<Halide::Argument> > const&, std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const&, std::__u::map<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> >, Halide::JITExtern, std::__u::less<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > >, std::__u::allocator<std::__u::pair<std::__u::basic_string<char, std::__u::char_traits<char>, std::__u::allocator<char> > const, Halide::JITExtern> > > const&, std::__u::vector<Halide::Internal::JITModule, std::__u::allocator<Halide::Internal::JITModule> > const&) third_party/halide/halide/src/WasmExecutor.cpp:2718:32
    #4 0x56275b47238c in Halide::Pipeline::compile_jit(Halide::Target const&) third_party/halide/halide/src/Pipeline.cpp:635:33
    #5 0x56275b4786d6 in Halide::Pipeline::realize(Halide::JITUserContext*, Halide::Pipeline::RealizationArg, Halide::Target const&, Halide::ParamMap const&) third_party/halide/halide/src/Pipeline.cpp:1130:5
    #6 0x56275b477724 in Halide::Pipeline::realize(Halide::JITUserContext*, std::__u::vector<int, std::__u::allocator<int> >, Halide::Target const&, Halide::ParamMap const&) third_party/halide/halide/src/Pipeline.cpp:757:9
    #7 0x56275b476e8e in Halide::Pipeline::realize(std::__u::vector<int, std::__u::allocator<int> >, Halide::Target const&, Halide::ParamMap const&) third_party/halide/halide/src/Pipeline.cpp:738:12
    #8 0x56275aba88d0 in Halide::Func::realize(std::__u::vector<int, std::__u::allocator<int> >, Halide::Target const&, Halide::ParamMap const&) third_party/halide/halide/src/Func.cpp:3276:23
    #9 0x56275ab14b3f in main third_party/halide/halide/test/correctness/align_bounds.cpp:52:32
    #10 0x7f4176b038d2 in __libc_start_main (/usr/grte/v5/lib64/libc.so.6+0x628d2) (BuildId: 7cfed7708e5ab7fcb286b373de21ee76)
    #11 0x56275aa47ae9 in _start /usr/grte/v5/debug-src/src/csu/../sysdeps/x86_64/start.S:120

SUMMARY: AddressSanitizer: heap-use-after-free third_party/wabt/src/interp/interp-inl.h:187:7 in wabt::interp::FreeList<wabt::interp::Object*>::~FreeList()
Shadow bytes around the buggy address:
  0x0c0c80064ac0: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x0c0c80064ad0: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c0c80064ae0: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c0c80064af0: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x0c0c80064b00: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
=>0x0c0c80064b10: fa fa fa fa[fd]fd fd fd fd fd fd fd fa fa fa fa
  0x0c0c80064b20: fd fd fd fd fd fd fd fa fa fa fa fa fd fd fd fd
  0x0c0c80064b30: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fa
  0x0c0c80064b40: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c0c80064b50: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x0c0c80064b60: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==5569==ABORTING
```

**Comment by zherczeg** [comment]
This is very interesting. I couldn't find a `halide` project in the `third_party` repository, but I suspect this is the project you are talking about: `https://github.com/halide`

The lines in the backtrace do not exactly match to the github repository, but maybe it is enough to figure out what is happening:

https://github.com/halide/Halide/blob/main/src/WasmExecutor.cpp#L2310
`store = wabt::interp::Store(calc_features(halide_module.target()));`

I suspect this is the `WasmExecutor.cpp:2312:13` line in the backtrace.

Then, in the same line the `Store` destructor runs with a different colum info: `WasmExecutor.cpp:2312:5`

Then the destructor of `Store` also runs again in `Halide::Internal::WasmModuleContents::~WasmModuleContents()` `WasmExecutor.cpp:2287:35`

The first `Store` destructor call is the strange thing. It looks like the store (and its content) is destroyed twice, and the first is during the assignment of `store =`. Do you have any idea why the destructor is called by this assignment? Something c++ related?

**Comment by zherczeg** [comment]
I have added a test to `test-interp.cc`

```
TEST_F(InterpGCTest, Store_Assignemnt) {
  wabt::interp::Store store;

  store = wabt::interp::Store();
  EXPECT_EQ(1u, store_.object_count());
}
```

This crashes, because the destructor is indeed called twice.

**Comment by sbc100** [comment]
Right, this is because in C++ `Foo f = Foo();` contructs two instances of Foo.. one on the RHS of the expression and one of the LHS which gets assigned to using `operator=`.

There are various options for how to handle such cases:

1. Support a normal operator= method that can make a copy of a store
2. Support a move assignment operator= that can move the contents from one instance to another: https://en.cppreference.com/w/cpp/language/move_assignment
3. Remove support for assigned/copying of stores completely.

Most likely option (2) is the best option since it won't break existing code.

CC: @steven-johnson

File: src/interp/interp-inl.h (hunk 1 above)
Comment (sbc100) [reply src/interp/interp-inl.h:189]: Can we switch back to `std::unique_ptr` and use `.get()` to grap its underlying pointer?

**Comment by zherczeg** [comment]
Sorry, it is not the Features object, it is the Null object which is destroyed twice:
```
Ref ref{objects_.New(new Object(ObjectKind::Null))};
```

It looks like the FreeList data is copied, then the temporary `Store` is deleted. Perhaps a copy operator is needed for FreeList?

**Comment by sbc100** [comment]
I think the new approach means we can't use `std::unique_ptr` .. which means we need to handler the assignment and move assignment operator ourselves.

I suggest I suggest we delete the copy assignment operator (the one that copies rather than moves) completely form both `FreeList` and from `Store`.   For the move assignment operator I think it probably makes sense to delete that too.. although it would require an update to the halide code I think.    Implementing a working move assignment operator might not be worth the effort.

**Comment by zherczeg** [comment]
An attempt to fix this: #1854

**Comment by zherczeg** [comment]
Deleting also an option. I can do that as well.

**Comment by sbc100** [comment]
I think #1854 should work.

**Pull request merged**